### ISSUE TITLE
BrainScriptNetworkBuilder: The training criterion is not a number (NAN).

### ISSUE THREAD

**Issue by Henry-White** [Microsoft/CNTK#607]
Latest build. Runs on CPU. Using MKL (ACML disabled).
I have experienced this, but with the Simple Network Builder. I got this error simply by reducing the value to 10 (minibatchSize = 10). Reattached (minibatchSize = 30) and the error has disappeared.
But with BrainScriptNetworkBuilder this method does not works.

Starting minibatch loop.
 Epoch[ 1 of 30]-Minibatch[   1- 500]: **ce = 1.#QNAN000** * 5000; errs = 0.000% * 5000; time = 1.6997s; samplesPerSecond = 2941.6

[CALL STACK]
    > Microsoft::MSR::CNTK::SGD<float>::  TrainOrAdaptModel
    - Microsoft::MSR::CNTK::SGD<float>::  Train
    - DoTrain<Microsoft::MSR::CNTK::ConfigParameters,float>  
    - DoCommands<float>  
    - wmainOldCNTKConfig
    - wmain1
    - wmain
    - __tmainCRTStartup
    - BaseThreadInitThunk
    - RtlUserThreadStart

EXCEPTION occurred: The training criterion is not a number (NAN).

**Comment by Henry-White** [comment]
Interesting... Decreased minibatchSize = 2 and moved forward at one step:
 Epoch[ 1 of 30]-Minibatch[   1- 500]: ce = -0.00000000 * 500; errs = 0.000% * 500; time = 1.6253s; samplesPerSecond = 307.6
 Epoch[ 1 of 30]-Minibatch[ 501-1000]: ce = 1.#QNAN000 * 500; errs = 0.000% * 500; time = 0.9243s; samplesPerSecond = 541.0

**Comment by frankseide** [comment]
Does this setup specify learningRatesPerMB? In that case, reducing the MB size will increase your effective learning rate per sample, which may cause divergence.

Although the ce=0 and errs=0 here indicate some other problem.

Could you summarize which configs (MKL, ACML, GPU, BS) work and which don't?

**Comment by Henry-White** [comment]
I had tried different ways: learningRatesPerMB = 1.5*10:0.7*10:0.3 and learningRatesPerSample = 0.09*10:0.03*10:0.01 and many others variants with values and parameters. No result.

> Could you summarize which configs (MKL, ACML, GPU, BS) work and which don't?

Actually, non of them shows signs of learning.
BS could not test due to an exeption

I could suppose that something wrong with my data, but on the same data set I get pretty well convergence with another lib by using a standard and resilient backpropagation ( there is a problem with oscillation and over-training but that is another story). Data has normal distribution and very well normalized (1;-1). I deliberately made a simple configuration to get at least some data, but didn't get. I put it here just for check.
`RootDir = ".."
ConfigDir = "$RootDir$/Config"
DataDir   = "$RootDir$/Data"
OutputDir = "$RootDir$/Output"
ModelDir = "$OutputDir$/Models"
deviceId = -1

command = Train:Test:Write

precision = "float"
initOnCPUOnly=true
traceLevel = 1
modelPath = "$ModelDir$/simple.dnn"
initValueScale = 0.6

Train = [
    action = "train"

    BrainScriptNetworkBuilder = [
	SDim = 100 # feature dimension
	HDim = 30   # hidden dimension
	LDim = 1    # number of classes

	features = Input (SDim, tag='feature') 
	labels   = Input (LDim, tag='label') 

	W0 = Parameter (HDim, SDim) ; b0 = Parameter (HDim, 1)  
	W1 = Parameter (LDim, HDim) ; b1 = Parameter (LDim, 1)

	r = RectifiedLinear (W0 * features + b0)
	z = W1 * r + b1 

	P    = Softmax (z, tag='output')
	ce   = CrossEntropyWithSoftmax (labels, z, tag='criterion')
	errs = ErrorPrediction         (labels, z, tag='evaluation')
//	outputNodes = (z:P)
    ]

    SGD = [
        epochSize = 0
        minibatchSize = 2
        learningRatesPerMB = 1.5*10:0.7*10:0.3
//	learningRatesPerSample = 0.09*10:0.03*10:0.01
//        momentumPerMB = 0.9*20:0.1
        momentumPerMB = 0.9
//        L2RegWeight = 0.03
        dropoutRate = 0.3*10:0.2*10:0.03
        maxEpochs = 30
    ]
    
    reader = [
        readerType = "CNTKTextFormatReader"
        file = "$DataDir$/36-0-100-Dt_DIFF-HP6,HPL,48_8-Pf1000-Nrm6,6_pr(16)_CNTK.dat"
        input = [
            features = [
		alias = "In"
		dim = 100      # two-dimensional input data
                format = "dense"
            ]
            labels = [
		alias = "Out"
		dim = 1        # One label dimension
                format = "dense"
            ]
        ]
    ]    
]

Test = [
    action = "test"
    minibatchSize = 10

    reader = [
        readerType = "CNTKTextFormatReader"
        file = "$DataDir$/36-0-100-Dt_DIFF-HP6,HPL,48_8-TEST-Pf1000-Nrm6,6_pr(16)_CNTK.dat"
        input = [
            features = [
		alias = "In"
		dim = 100      # two-dimensional input data
                format = "dense"
            ]
            labels = [
		alias = "Out"
		dim = 1        # One label dimension
		format = "dense"
            ]
        ]
    ]    
]

Write = [
    action = "write"
    reader=[
        readerType = "CNTKTextFormatReader"
	file = "$DataDir$/36-0-100-Dt_DIFF-HP6,HPL,48_8-TEST-Pf1000-Nrm6,6_pr(16)_CNTK.dat"
	input = [
          features = [
		alias = "In"
		dim = 100      # two-dimensional input data
		format = "dense"
		]
  
          labels = [
		alias = "Out"
		dim = 1        # One label dimension
		format = "dense"
		]
	]
    ]
//    outputNodeNames = PosteriorProb : labels
    outputNodeNames = OutputNodes : z : P : ce : errs
    outputPath = "$OutputDir$/SimpleOutput" 
]`

**Comment by Henry-White** [comment]
Under Debug:
Starting Epoch 1: learning rate per sample = 0.090000  effective momentum = 0.900000  momentum as time constant = 94.9 samples
BlockRandomizer::StartEpoch: epoch 0: frames [0..43309] (first sequence at sample 0), data subset 0 of 1
Starting minibatch loop.
HasNan: NaN detected at TrainOneEpoch/UpdateWeights():  (0,0) in (30,100) matrix
About to throw exception 'W0 LearnableParameter operation has NaNs in functionValues after parameter update.'

Continuing to explore...

**Comment by frankseide** [comment]
I noticed this:

    LDim = 1    # number of classes

followed by a Softmax. Softmax is normally used to classify one out of N classes. If the number of classes is 1, it will always output 1.

What is it that you are trying to do? Are you trying to predict values that are either 0 or 1? For that, you would use the Logistic loss (`Logistic (labels, z, tag='criterion')`).

**Comment by Henry-White** [comment]
As always, thank you very much for your help. I'm very appreciated for this.

I thought "softmax" at the end of the BS declaration is a softmax (tanh) normalization for output stabilization. Lack of documentation and experience in DNN is forcing me make stupid errors.

About, what we are trying to do... We have some x-ray spectrometry data. Input and output vectors are float with symmetrical distribution (-1, 1). Dims are 100 and 1. Actually we were trying to use RPROP-networks for extracting main harmonics. Because of high concentration of highly correlation neurons on start of the input vector we are using a slight prediction shift. Due to some inherent problems with RPROP networks, now, we are trying to use CNTK for this task.

So, the actual question is: 
Does CNTK symmetric functions implementation or all of them has input values in the range (0, 1) ?

**Comment by frankseide** [comment]
You are most welcome.

I am not entirely sure if I understand the harmonics bit, but could you use square error with an actual tanh? E.g.:

    z = ...  // your top layer as before
    pred = TanH (z)   // predicted value
    err = SquareError (pred, target, tag='criterion')   // target = desired output value -1..1

**Comment by Henry-White** [comment]
Many thanks for your reply. I got quite true result for 30 Epochs.
I think that this thread does not seems to be an issue any more.
Close.
